rpc/src/v1/impls/eth.rs

```diff
@@ -356,113 +356,144 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM> Eth for EthClient<C, SN, S, M, EM> where
 		Ok(RpcU256::from(take_weak!(self.client).chain_info().best_block_number))
 	}
 
-	fn balance(&self, address: RpcH160, num: Trailing<BlockNumber>) -> Result<RpcU256, Error> {
+	fn balance(&self, address: RpcH160, num: Trailing<BlockNumber>) -> BoxFuture<RpcU256, Error> {
 		let address = address.into();
-		match num.0 {
-			BlockNumber::Pending => Ok(take_weak!(self.miner).balance(&*take_weak!(self.client), &address).into()),
-			id => {
-				let client = take_weak!(self.client);
 
-				check_known(&*client, id.clone())?;
-				match client.balance(&address, id.into()) {
-					Some(balance) => Ok(balance.into()),
-					None => Err(errors::state_pruned()),
+		let inner = || {
+			match num.0.clone() {
+				BlockNumber::Pending => Ok(take_weak!(self.miner).balance(&*take_weak!(self.client), &address).into()),
+				id => {
+					let client = take_weak!(self.client);
+
+					check_known(&*client, id.clone())?;
+					match client.balance(&address, id.into()) {
+						Some(balance) => Ok(balance.into()),
+						None => Err(errors::state_pruned()),
+					}
 				}
 			}
-		}
+		};
+
+		future::done(inner()).boxed()
```

Comment:
Yes, might be worth doing. I don't like to hide the allocation but I suppose it's evident in the return type.